### PR TITLE
CRM-20260, fixed code for retrieving last financial item for contribu…

### PR DESCRIPTION
**Pull request by pradpnayak**
…tion

----------------------------------------

  https://issues.civicrm.org/jira/browse/CRM-20260

### CODE REVIEW

**Comment by pradpnayak** [comment]
depends on https://github.com/civicrm/civicrm-core/pull/9590

**Comment by pradpnayak** [comment]
jenkins, retest this please

**Comment by coldrunKacper** [comment]
Looks fine to me. Approved.

**Comment by pradpnayak** [comment]
Can this be merged?

**Review by eileenmcnaughton — CHANGES_REQUESTED** [review]
OK, so here is my problem with this one
1) I don't know what problem it's trying to fix, it doesn't have it's own JIRA ticket & there is nothing to explain how to replicate any problem
2) I don't know what was reviewed - code only , or the actual bug.

OTOH I agree the code looks good & the test coverage is good.

Can we get the administrative side of this tidied up.

**Comment by eileenmcnaughton** [comment]
per other comments, setting to WIP to get a clear ticket that this relates to & steps to replicate

**Comment by JoeMurray** [comment]
Pradeep will be updating the administrative side. I think it would be good for there to be an indication that testing has actually worked once the issue description of problem is provided.

**Comment by pradpnayak** [comment]
Created ticket at https://issues.civicrm.org/jira/browse/CRM-20260 and added steps to replicate

File: CRM/Financial/BAO/FinancialItem.php [hunk 1]
```diff
@@ -295,13 +295,28 @@ public static function checkContactPresent($contactIds, &$error) {
    * @return object CRM_Core_DAO
    */
   public static function getPreviousFinancialItem($entityId, $entityTable = 'civicrm_line_item') {
+    $liabilityAccountType = CRM_Core_OptionGroup::getValue('financial_account_type', 'Liability', 'name');
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: You need to provide some code comments here to explain what you are doing with Liability Type financial accounts.

File: CRM/Financial/BAO/FinancialItem.php [hunk 2]
```diff
@@ -295,13 +295,28 @@ public static function checkContactPresent($contactIds, &$error) {
...
     $queryParams = array(
       1 => array($entityId, 'Integer'),
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: throughout system entity_table always is entered in schema before entity_id. Stylistically would be nice to keep that order.

File: CRM/Financial/BAO/FinancialItem.php [hunk 3]
```diff
@@ -295,13 +295,28 @@ public static function checkContactPresent($contactIds, &$error) {
...
+    $query = 'SELECT cfi.id, cfi.description, cfi.amount, cfi.financial_account_id, cfi.status_id
+      FROM `civicrm_financial_item` cfi
+        INNER JOIN civicrm_entity_financial_trxn ceft 
+          ON ceft.entity_id = cfi.id AND ceft.entity_table = "civicrm_financial_item"
+        INNER JOIN civicrm_financial_trxn cft ON cft.id = ceft.financial_trxn_id
+        INNER JOIN civicrm_financial_account cfa ON cfa.id = cfi.financial_account_id
+      WHERE cfi.entity_table = %2 AND cfi.entity_id = %1 AND 
+        CASE
+          WHEN cfa.financial_account_type_id = %3
+            AND cfi.financial_account_id  = from_financial_account_id
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: You should put table alias or table name before from_financial_account_id.

File: CRM/Financial/BAO/FinancialItem.php [hunk 4]
```diff
@@ -295,13 +295,28 @@ public static function checkContactPresent($contactIds, &$error) {
...
+          THEN 1
+          WHEN cfa.financial_account_type_id = %3
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: Why not simplify by replacing CASE with ((cfa.financial_account_type_id = %3 AND cfi.financial_account_id  = from_financial_account_id) OR financial_account_type_id <> %3)

File: CRM/Financial/BAO/FinancialItem.php [hunk 5]
```diff
@@ -295,13 +295,28 @@ public static function checkContactPresent($contactIds, &$error) {
...
+    $liabilityAccountType = CRM_Core_OptionGroup::getValue('financial_account_type', 'Liability', 'name');
     $queryParams = array(
       1 => array($entityId, 'Integer'),
       2 => array($entityTable, 'String'),
+      3 => array($liabilityAccountType, 'Integer'),
     );
-    $query = 'SELECT id, description, status_id, financial_account_id 
-      FROM civicrm_financial_item
-      WHERE entity_id = %1 AND entity_table = %2 ORDER BY id DESC LIMIT 1';
+    $query = 'SELECT cfi.id, cfi.description, cfi.amount, cfi.financial_account_id, cfi.status_id
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: I don't understand why you're only choosing these fields. Don't you need to copy contact_id and currency to new financial_item? What about *

File: CRM/Financial/BAO/FinancialItem.php (hunk 1 above)
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/FinancialItem.php]: Note that function is deprecated - if you use an IDE like php storm that leaps out at you.

screenshot 2017-03-15 20 31 07

I believe in more cases you can do things with 

CRM_Core_PseudoConstant::getLabel()
CRM_Core_PseudoConstant::getKey()
CRM_Core_Pseudoconstant::getName()

**Comment by eileenmcnaughton** [comment]
@pradpnayak with those instructions I was able to replicate the bug & confirm your changes fixed it.

However, the code that this touches on is pretty nasty & I wasn't able to verify it without going through the full process I would go through to fix the  bug - ie. writing a unit test that fails & then fixing the code until it passes. In the process I would up with an alternate fix. Although it looks longer
https://github.com/civicrm/civicrm-core/pull/9984 I think most of the extra length is in the full-flow-test and the fact I renamed a few variables & function names. The actual key function is shorter and I was able to remove the sql from it (ideally we would have a lot less sql in the codebase).

I tried to comment it really heavily in github so you could make sense of it & review it, let me know.

I also logged a follow on bug that I identified & linked to a fix from JIRA https://issues.civicrm.org/jira/browse/CRM-20276  - that fix depends on my PR & needs to be reviewed after this one is resolved

**Comment by monishdeb** [comment]
@JoeMurray @eileenmcnaughton I have made the desired changes requested in the PR and tested.

**Comment by colemanw** [comment]
@eileenmcnaughton do these latest changes look good to you?

**Comment by eileenmcnaughton** [comment]
@pradpnayak I pulled your test into #9984 to see if it would still pass with the simpler api-based method and it's passing! (note I changed it a little since you did your QA). Do you think there is still a scenario #9984 is not working with or should we merge that one (which has other improvements / additional tests  / less in-code sql)

**Comment by eileenmcnaughton** [comment]
I order to unblock this and the follow on bug fix I have merged #9984 and am closing this.

I understand that there is still a use-case that needs a test & which will alter the merged fix, that affects users of deferred_revenue under some config. I would not have merged the other had we not just cut the rc.  

This has been really stalled & hard to get clarity on so getting to a clear point on what has been merged seems like a priority to me

**Pull request closed without merging**